### PR TITLE
Fix Numpy compatibility and StorageObj attributes

### PR DESCRIPTION
**Pull request by polsm91**
Add compatibility with numpy. Some numpy versions use array_wrap, but its deprecated and the new process is to intercept Numpy operations is implementing  the __ufunc__ method.

The second problem is the StorageObj attributes were generated and keep in memory. (The __getattribute__ method wasn't implemented). So accessing the attributes produced an in memory access without intercepting the call.

### CODE REVIEW

File: hecuba/IStorage.py [hunk 1]
```diff
@@ -170,6 +173,20 @@ def _count_name_collision(self, attribute):
                                    [self._ksp])
         return len(filter(lambda (t_name, ): m.match(t_name), q))
 
+
+
+    def _build_istorage_obj(self, obj_info, so_name, storage_id):
```
Comment (cugni) [inline hecuba/IStorage.py]: this does not create a storageobj (not only at least). Change the name. Also make it static

File: hecuba/storageobj.py [hunk 2]
```diff
@@ -350,43 +348,56 @@ def make_persistent(self, name):
...
+            except AttributeError:
+                #Attribute unset, no action needed
+                pass
         self._store_meta(self._build_args)
 
     def stop_persistent(self):
         """
             The StorageObj stops being persistent, but keeps the information already stored in Cassandra
         """
+        for obj_name, obj_info in self._persistent_props.items():
```
Comment (cugni) [inline hecuba/storageobj.py]: check this: why are you using build in functions? Can't you just check if they are IStorage instances?

File: hecuba/storageobj.py [hunk 3]
```diff
@@ -350,43 +348,56 @@ def make_persistent(self, name):
...
-
-        if hasattr(self, '_persistent_dicts'):
-            for pers_dict in self._persistent_dicts:
-                pers_dict.delete_persistent()
 
-        if hasattr(self, '_storage_objs'):
-            for so in self._storage_objs:
-                so.delete_persistent()
+        for (attr_name,attr_info) in self._persistent_props.iteritems():
+            if attr_info['type'] not in IStorage._basic_types:
```
Comment (cugni) [inline hecuba/storageobj.py]: as before

File: hecuba/storageobj.py [hunk 4]
```diff
@@ -350,43 +348,56 @@ def make_persistent(self, name):
...
+                    pass
 
         query = "TRUNCATE TABLE %s.%s;" % (self._ksp, self._table)
         log.debug("DELETE PERSISTENT: %s", query)
         config.session.execute(query)
 
-    def __getattr__(self, attribute):
+        self._is_persistent = False
+
+    def __getattribute__(self, attribute):
```
Comment (cugni) [inline hecuba/storageobj.py]: you CANNOT override the __getattribute__ ! If you do you won't be able to access to any class method.

Comment (cugni) [reply hecuba/storageobj.py]: check the __getattr__

File: hecuba/hdict.py [hunk 5]
```diff
@@ -437,18 +431,23 @@ def make_persistent(self, name):
...
         # Storing all in-memory values to cassandra
         for key, value in dict.iteritems(self):
+            if issubclass(value.__class__, IStorage):
+                # new name as ksp.table_valuename, where valuename is either defined by the user or set by hecuba
+                val_name = self._ksp + '.' + self._table + '_' + self._columns[0][0]
+                value.make_persistent(val_name)
+                value = value._storage_id
             self._hcache.put_row(self._make_key(key), self._make_value(value))
         if hasattr(self, '_indexed_args') and self._indexed_args is not None:
             index_query = 'CREATE CUSTOM INDEX IF NOT EXISTS ' + self._table + '_idx ON '
```
Comment (cugni) [inline hecuba/hdict.py]: add a self.__dict__ = {} to clean the memory dict after make the StorageDict persistent

Comment (cugni) [reply hecuba/hdict.py]: add tests to check it work (try with a storageObj with a custom function)

File: hecuba/storageobj.py (hunk 2 above)
Comment (polsm91) [reply hecuba/storageobj.py]: We only need to stop the objects which are in memory, stopping something that is in Cassandra doesn't do anything. Stop_persistent only switches from persistent to non persistent (without deleting anything).
About the second point, yes, it's the same. In this way getattribute is called fewer times if we check the type before. However, it does a search in a list which isn't efficient too. I will change it.

**Pull request merged**